webapp/src/main/java/com/box/l10n/mojito/service/machinetranslation/MachineTranslationService.java

```diff
@@ -0,0 +1,165 @@
+
+        for (int i = 0; i < textSources.size(); i++) {
+            String sourceText = textSources.get(i);
+            TextUnitTranslationGroup textUnitTranslationGroupDTO = translationsResponseDTO.getTextUnitTranslations().get(i);
+
+            // Note: to further increase leverage performance at a later date, we should introduce a bulk leveraging
+            // function that will take in a list of source strings, a list of target locales and will return one single
+            // result for each source string and locale combination.
+            List<TextUnitDTO> textUnitDTOsForLeveraging = getLeveragedTextUnitDTOS(sourceText);
+
+            if (!textUnitDTOsForLeveraging.isEmpty()) {
+                Comparator<TextUnitDTO> tmTextUnitCreatedDateComparator = Comparator.comparing(TextUnitDTO::getTmTextUnitCreatedDate);
+
+                // Pick the most recent created translations for each language
+                Map<String, TextUnitDTO> mostRecentTranslationsByLanguage = textUnitDTOsForLeveraging.stream()
+                        .collect(Collectors.toMap(TextUnitDTO::getTargetLocale, Function.identity(),
+                                BinaryOperator.maxBy(tmTextUnitCreatedDateComparator)));
+
+                // Copy translations from leveraged matches
+                for (int j = 0; j < textUnitTranslationGroupDTO.getTranslations().size(); j++) {
+                    Translation translation = textUnitTranslationGroupDTO.getTranslations().get(j);
+
+                    // Note: the leveraging mode is locale code specific. If the translation requests ask for "pt" but
+                    // we only have a "pt-BR" translation, or the other way around, we do not use the leveraged match.
+                    if (mostRecentTranslationsByLanguage.containsKey(translation.getLocaleTag())) {
+                        TextUnitDTO matchedLanguageTextUnit = mostRecentTranslationsByLanguage.get(translation.getLocaleTag());
+
+                        translation.setText(matchedLanguageTextUnit.getTarget());
+                        translation.setMatchedTextUnitId(matchedLanguageTextUnit.getTmTextUnitId());
+                        translation.setMatchedTextUnitVariantId(matchedLanguageTextUnit.getTmTextUnitVariantId());
+                        translation.setTranslationSource(TranslationSource.MOJITO_TM_LEVERAGE);
+                    }
+                }
+            } else {
+                logger.debug("No TMTextUnit Match found for content: {}", sourceText);
+            }
+        }
+        return translationsResponseDTO;
+    }
+
+    @SuppressWarnings("SpringCacheableComponentsInspection")
+    @Cacheable("machineTranslation")
+    List<TextUnitDTO> getLeveragedTextUnitDTOS(String sourceText) {
+        TMTextUnit searchTmTextUnit = new TMTextUnit();
+        searchTmTextUnit.setContent(sourceText);
+
+        return leveragerByContent.getLeveragingMatches(searchTmTextUnit, null, null);
```

Comment:
is that fine in term of performance?

Reply:
Performance isn't great as that API returns more results than desired generally and the query isn't solely optimized for this purpose, which is why I've included a skipLeveraging parameter - though it is usable if it's not on a critical path - and especially fast on subsequent calls with the in-memory caching being enabled on getLeveragedTextUnitDTOS().

Ideally we'd do this operation in bulk for multiple sources and only return the latest/best match per language, as per the comment from line 122.

Reply:
I'll add repositoryId as an additional parameter to be able to narrow the search - if not specified, we'll search in the entire DB. 

Anything that narrows the scope of the search is welcome as far as I'm concerned.